### PR TITLE
Create abstraction for reversing `<iframe>`

### PR DESCRIPTION
**Pull request by danielbachhuber**
This will help prevent duplicated regex everywhere

See #99

### CODE REVIEW

**Comment by danielbachhuber** [comment]
@goldenapples Thoughts? #reviewmerge and I can port the remaining over?

**Comment by danielbachhuber** [comment]
Actually, want to check one more thing.

File: inc/shortcodes/class-shortcode.php [hunk 1]
```diff
@@ -55,4 +55,54 @@ public static function callback( $attrs, $content = '' ) {
...
+	 * @param string $content
+	 * @return array|false
+	 */
+	protected static function parse_iframes( $content ) {
+
+		if ( false === stripos( $content, '<iframe' ) ) {
+			return false;
+		}
+
+		if ( preg_match_all( '#(.+\r?\n?)?(<iframe([^>]+)>[^<]{0,}</iframe>)(\r?\n?.+)?#', $content, $matches ) ) {
```
Comment (danielbachhuber) [inline inc/shortcodes/class-shortcode.php]: @goldenapples any ideas why this regex would fail on PHP 5.3?

Comment (goldenapples) [inline inc/shortcodes/class-shortcode.php]: I'm not sure what you're trying to do with the beginning and end delimiters. `(.+\r?\n?)?` Is that just supposed to swallow the previous and next lines of the content if any exist?

**Comment by danielbachhuber** [comment]
Hm. To debug:

1. It only fails for Giphy and Vimeo, which use heredoc to generate the test strings. `Test_Shortcode` passes correctly.
2. _Something_ does work, because the problem is that the replacement value is invalid.

File: inc/shortcodes/class-shortcode.php (hunk 1 above)
Comment (danielbachhuber) [inline inc/shortcodes/class-shortcode.php]: > Is that just supposed to swallow the previous and next lines of the content if any exist?

I'm not sure "swallow" is the right word. It adds them to the `iframe` object, to make them available in replacement as needed.

**Comment by danielbachhuber** [comment]
Oh, I know what's happening: `parse_url()` failure.

**Comment by goldenapples** [comment]
Question: what if `parse_iframes` returned an array of attributes, so that each individual shortcode didn't have to do its own regex parsing?

Running `shortcode_parse_atts()` over the matched attributes string of the `<iframe>` tag might be enough in most cases. Setting up an actual DOM Document element might be necessary in some cases...

**Comment by danielbachhuber** [comment]
> Question: what if parse_iframes returned an array of attributes, so that each individual shortcode didn't have to do its own regex parsing?

It does?

If the build passes, this is ready for #reviewmerge

File: tests/test-shortcode.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+		$iframe_obj = new stdClass;
+		$iframe_obj->original = '<iframe src="//giphy.com/embed/ihfrhIgdkQ83C" width="480" height="293" allowFullScreen></iframe>';
+		$iframe_obj->before = '';
+		$iframe_obj->after = '<p><a href="http://giphy.com/gifs/jtvedit-jtv-rogelio-de-la-vega-ihfrhIgdkQ83C">via GIPHY</a></p>';
+		$iframe_obj->inner = '';
+		$iframe_obj->attrs = array(
+			'src'             => '//giphy.com/embed/ihfrhIgdkQ83C',
+			'width'           => '480',
+			'height'          => '293',
+			'allowFullScreen' => null,
```
Comment (danielbachhuber) [inline tests/test-shortcode.php]: @goldenapples These are parsed attributes

**Comment by danielbachhuber** [comment]
:boom:

**Comment by goldenapples** [comment]
> It does?

/facepalm Sorry, I must have been looking at something else altogether.

File: inc/shortcodes/class-shortcode.php [hunk 3]
```diff
@@ -55,4 +55,61 @@ public static function callback( $attrs, $content = '' ) {
...
+		if ( preg_match_all( '#(.+\r?\n?)?(<iframe([^>]+)>([^<]+)?</iframe>)(\r?\n?.+)?#', $content, $matches ) ) {
+			$iframes = array();
+			foreach ( $matches[0] as $key => $value ) {
+				$iframe = new \stdClass;
+				$iframe->original = $matches[2][ $key ];
+				$iframe->before = $matches[1][ $key ];
+				$iframe->attrs = array( 'src' => '' );
+				$iframe->inner = $matches[4][ $key ];
+				$iframe->after = $matches[5][ $key ];
+				$parts = explode( ' ', $matches[3][ $key ] );
```
Comment (goldenapples) [inline inc/shortcodes/class-shortcode.php]: What does this do if a space appears inside an attribute value, like a list of class names or a title?

Comment (danielbachhuber) [inline inc/shortcodes/class-shortcode.php]: Ah, probably breaks. Want to write a test case for it, and commit a fix? I'm about to sign off for an hour.

Comment (goldenapples) [inline inc/shortcodes/class-shortcode.php]: On it.

**Pull request merged**